### PR TITLE
radicle: namespace peers

### PR DESCRIPTION
**Pull request by FintanH**
The previous iteration of the storage layout for projects organised peer's subtrees by the `refs/remotes` category. This layout leads to problems when linking a working copy to the storage project. The fetch refspec would look something like:

  `fetch = +refs/remotes/abc/heads/*:refs/remotes/rad/*`

This works perfectly fine when running:

  git fetch rad

but will fail when one tries to specify:

  `git fetch rad main`

Instead, the command would need to be:

  `git fetch remotes/abc/heads/main`

This patch proposes to move the use of `refs/namespaces` for each peer. The layout looks similar in that each peer has a subtree, but instead it is under `refs/namespaces/<peer>/refs/...`.

The previous refspec would look more familiar in its new form:

  `fetch = +refs/heads/*:refs/remotes/rad/*`

With this in place running:

  `git --namespace=abc rad main`

Will work as expected and place the reference under `refs/remotes/rad/main`. Similary, `git --namespace=abc push` will work as expected.

Another benefit of this approach is that the use of namespacing means that a tag is pushed it will be placed under the respective namespace rather than in the global `refs/tags` category.

The actual implementation needed to work around the fact that `git2`/`libgit2` does not have any options for fetching or pushing using namespaces. The working copy code uses processes to shell out to `git` for namespacing needs.

### CODE REVIEW

File: radicle/src/storage.rs
Comment (FintanH) [inline radicle/src/storage.rs:179]: I'm curious about this :eyes:

File: radicle/src/storage.rs
Comment (FintanH) [inline radicle/src/storage.rs:189]: Also, this looks like a leaky function, but a second pair of eyes might help confirm that :)

File: radicle/src/git.rs
Comment (cloudhead) [inline radicle/src/git.rs:19]: This diff line reminds me of when Kim would patch my Haskell code but first he would import half of Control.Applicative

File: radicle/src/git.rs
Comment (cloudhead) [inline radicle/src/git.rs:135]: Meta: I find this API or naming strange: it sounds like it is namespacing `input`, ie. "give me a namespaced version of input", even though input is already namespaced. I think perhaps this should be called `as_namespaced()` or `Namespaced::try_from(input)`

File: radicle/src/git.rs
Comment (cloudhead) [inline radicle/src/git.rs:268]: Maybe use the `run` function below? Might need to be changed to accept a namespace param?

File: radicle/src/git.rs
Comment (cloudhead) [inline radicle/src/git.rs:245]: I think this is a good idea, I'm not a fan of optional parameters in rust using `Option`; we could take a slice of pushspecs instead :shrug:

File: radicle/src/storage.rs
Comment (cloudhead) [reply radicle/src/storage.rs:179]: This basically says that this remote belongs to a delegate of the project, ie. the public key is part of the project delegation. Does that make sense?

File: radicle/src/storage.rs
Comment (cloudhead) [inline radicle/src/storage.rs:263]: Thanks for adding docs!

File: radicle/src/storage/refs.rs
Comment (cloudhead) [inline radicle/src/storage/refs.rs:27]: WTB `git::qualified!`

File: radicle/src/storage/refs.rs
Comment (cloudhead) [inline radicle/src/storage/refs.rs:291]: Meta: another function name I think could be improved: `add` sounds like it's adding it to the end, I'd propose `with_namespace(remote)` or `namespaced(remote)`

File: radicle/src/git.rs
Comment (cloudhead) [inline radicle/src/git.rs:224]: I tend to not like having external URLs in repos because I hate broken/mutable links :grimacing:

**Comment by cloudhead** [comment]
This is great, thanks. Mainly minor points. I am realizing though that you may have some conflicts with master due to the unwrap-ifying, but I can help with that.

File: radicle/src/git.rs
Comment (FintanH) [reply radicle/src/git.rs:224]: Fair enough, I can create an issue instead to keep track of it.

File: radicle/src/storage.rs
Comment (FintanH) [reply radicle/src/storage.rs:179]: Ah, I think the comment threw me off with the use of "a project", but this `Remote` struct is just for a single project?

File: radicle/src/git.rs
Comment (FintanH) [reply radicle/src/git.rs:19]: lol, that's how you know it's a good abstraction -- everything is immensely useful ;P

I'm not the biggest fan of this re-exporting style. The rust-analyzer shortcut for importing will always import from the crate itself, and I never knew what was re-exported from here or not.

**Comment by FintanH** [comment]
> This is great, thanks. Mainly minor points. I am realizing though that you may have some conflicts with master due to the unwrap-ifying, but I can help with that.

No worries, I fixed everything up and I'll make issues for the namespace stuff in radicle-git!

File: radicle/src/storage.rs
Comment (cloudhead) [reply radicle/src/storage.rs]: Yeah, it's just for a single project. The refs it contains are scoped to a project/repo.

File: radicle/src/git.rs
Comment (cloudhead) [reply radicle/src/git.rs:21]: Yeah I mean I have it imported as `fmt` above, so you could also do `fmt::lit`, `fmt::name` etc.

File: radicle/src/storage.rs
Comment (FintanH) [reply radicle/src/storage.rs]: Cool! I updated the doc string :)

**Review by cloudhead — APPROVED** [review]

File: radicle/src/git.rs
Comment (FintanH) [reply radicle/src/git.rs:21]: Right, is that the standard to go by then or could I convince you to just import the crate where needed :eyes:? Would you want me to fix that up before merging?

**Pull request merged**
